### ISSUE TITLE
grey_allOf([grey_selected(), grey_accessibilityID("id")]) matching element that is not selected.

### ISSUE THREAD

**Issue by jasonm23** [google/EarlGrey#634]
using `grey_allOf` with `grey_selected` and `grey_accessibilityID` as the matchers, provides false positive matches.

The correct outcome can be had by using:

```
EarlGrey.select(elementWithMatcher: grey_accessibilityID("mySelectedElement")).assert(with: grey_selected())
```
or 

```
EarlGrey.select(elementWithMatcher: .assert(with: grey_selected()).assert(with: grey_accessibilityLabel("mySelectedElement"))
```

**Comment by tirodkar** [comment]
Do you have a repro case for this? The only way the issue should be seen if you have multiple elements with the same accessibilityID.

**Comment by jasonm23** [comment]
It's a unique accessibilityID.  Unfortunately the use case is within a client's proprietary codebase.

**Comment by jasonm23** [comment]
To clarify:

```
EarlGrey.select(elementWithMatcher: grey_allOf([ grey_accessibilityID("mySelectedElement"), grey_selected()])
```

Is giving the false match - tested against 10.1thru 10.3

**Comment by jasonm23** [comment]
(also tested against 11.0)

**Comment by tirodkar** [comment]
I see. Yeah, we'll test it out and

**Comment by jasonm23** [comment]
Grepped the output of `GREYElementHierarchy.hierarchyStringForAllUIWindows` the accessibilityID s we've been matching against, all occur exactly once.

**Comment by jasonm23** [comment]
e.g. after copying out the string to the clipboard...

```
pbpaste | grep journey
  |  |  |  |  |  |  |  |  |--<UIButton:0x7fd1d9f7dee0; AX=Y; AX.id='journey'; AX.label='Journey'; AX.frame={{10, 74}, {197, 50}}; AX.activationPoint={108.5, 99}; AX.traits='UIAccessibilityTraitButton'; AX.focused='N'; frame={{0, 0}, {197, 50}}; alpha=1>
```

**Comment by khandpur** [comment]
If accessibilityID is unique, you can just use grey_accessibilityID directly. I am a bit confused why you need to use grey_allOf and combine it with grey_selected.

On the other hand, we can add a test for `grey_allOf` with `grey_accessibilityID` and `grey_selected` so we have some coverage but without a concrete example showing how it's misbehaving, it's really hard for us to imagine what might be causing this issue.

**Comment by khandpur** [comment]
Also, could you make sure that the incorrect element being matched isn't a UI control and it doesn't have it doesn't have an [isSelected](https://developer.apple.com/documentation/uikit/uicontrol/1618203-selected?language=objc) property that returns YES?

**Comment by jasonm23** [comment]
Hi @khandpur - as outlined in the original post, we're getting a successful match using alternative syntax, using assert with:

The issue is to address the fact that `grey_allOf` is returning a false positive.

Shouldn't `grey_allOf` properly match the outcome of all selectors it contains by definition?

**Comment by tirodkar** [comment]
@jasonm23, to clarify - `grey_allOf()` returns the intersection of all the matchers provided to it.

Therefore, `grey_allOf(grey_text(@"Foo"), grey_kindOfClass([Foo class]), grey_interactable(), nil)` will return elements that have their text value as `Foo` and are of class `Foo` and are interactable.

What is the interaction i.e. the `performAction` or `assertWithMatcher:` that you are providing with `EarlGrey.select(elementWithMatcher: grey_allOf([ grey_accessibilityID("mySelectedElement"), grey_selected()])`? My guess is that the interaction is the one that is failing.

**Comment by jasonm23** [comment]
Hi @tirodkar 

The failing selector is:

```
EarlGrey.select(
    elementWithMatcher: grey_allOf([
        grey_selected(), 
        grey_accessibilityID("myUniqueAXid")
    ])
    .assert(with: grey_notNil())
```
